src/lib/optimizer/join_graph.cpp

```diff
@@ -0,0 +1,211 @@
+#include "join_graph.hpp"
+
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "optimizer/abstract_syntax_tree/abstract_ast_node.hpp"
+#include "optimizer/abstract_syntax_tree/join_node.hpp"
+#include "optimizer/abstract_syntax_tree/predicate_node.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+#include "utils/type_utils.hpp"
+
+namespace opossum {
+
+JoinEdge::JoinEdge(const std::pair<JoinVertexID, JoinVertexID>& vertex_indices,
+                   const std::pair<ColumnID, ColumnID>& column_ids, JoinMode join_mode, ScanType scan_type)
+    : vertex_indices(vertex_indices), column_ids(column_ids), join_mode(join_mode), scan_type(scan_type) {
+  DebugAssert(join_mode == JoinMode::Inner, "This constructor only supports inner join edges");
+}
+
+JoinEdge::JoinEdge(const std::pair<JoinVertexID, JoinVertexID>& vertex_indices, JoinMode join_mode)
+    : vertex_indices(vertex_indices), join_mode(join_mode) {
+  DebugAssert(join_mode == JoinMode::Cross, "This constructor only supports cross join edges");
+}
+
+JoinVertexPredicate::JoinVertexPredicate(ColumnID column_id, ScanType scan_type, const AllParameterVariant& value)
+    : column_id(column_id), scan_type(scan_type), value(value) {}
+
+JoinVertex::JoinVertex(const std::shared_ptr<AbstractASTNode>& node) : node(node) {}
+
+std::shared_ptr<JoinGraph> JoinGraph::build_join_graph(const std::shared_ptr<AbstractASTNode>& root) {
+  JoinGraph::Vertices vertices;
+  JoinGraph::Edges edges;
+
+  _traverse_ast_for_join_graph(root, vertices, edges);
+
+  return std::make_shared<JoinGraph>(std::move(vertices), std::move(edges));
+}
+
+JoinGraph::JoinGraph(Vertices&& vertices, Edges&& edges) : _vertices(std::move(vertices)), _edges(std::move(edges)) {}
+
+const JoinGraph::Vertices& JoinGraph::vertices() const { return _vertices; }
+
+const JoinGraph::Edges& JoinGraph::edges() const { return _edges; }
+
+void JoinGraph::print(std::ostream& out) const {
+  out << "==== JoinGraph ====" << std::endl;
+  out << "==== Vertices ====" << std::endl;
+  for (size_t vertex_idx = 0; vertex_idx < _vertices.size(); ++vertex_idx) {
```

Comment:
Since you added `JoinVertexID` as data type, we should either use it here or use `auto`